### ISSUE TITLE
Performance regression when using @torch.compile compared to no compilation

### ISSUE THREAD

**Issue by vladkvit** [pytorch/pytorch#144822]
### 🐛 Describe the bug

I was playing around with torch.compile functionality, and came across a ~20x slowdown when running this toy code (counts even numbers):

```
import torch
from datetime import datetime

batch_size = 1024 * 8000
final_num = 2147483647
num_batches = final_num // batch_size

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("device", device)


# @torch.compile
def bench():
    final_total = torch.tensor(0, dtype=torch.int32, device=device)

    for i in range(num_batches):
        batch = torch.arange(
            i * batch_size, (i + 1) * batch_size, dtype=torch.int32, device=device
        )
        batch_sum = torch.sum(batch % 2)
        final_total += batch_sum

    return final_total


a = datetime.now()
final_total = bench()
b = datetime.now()

print(f"Final Total: {final_total.item()}, {b-a} seconds")
```

On my 4090, I am seeing 0.13 seconds without the decorator and 2.3 seconds with the decorator (tested by running multiple times; excluding the very first run when the compilation happens). Is this expected behavior?

### Versions

Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: 19.1.4
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: 12.5.40
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 555.85
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: AMD Ryzen 7 7800X3D 8-Core Processor
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 4201
MaxClockSpeed: 4201
L2CacheSize: 8192
L2CacheSpeed: None
Revision: 24834

Versions of relevant libraries:
[pip3] flake8==7.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.16.0
[pip3] onnxruntime==1.17.3
[pip3] pytorch-ignite==0.5.1
[pip3] pytorch-lightning==2.1.4
[pip3] pytorch-metric-learning==2.3.0
[pip3] rapidocr-onnxruntime==1.3.24
[pip3] torch==2.5.1+cu124
[pip3] torch_geometric==2.4.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchao==0.7.0+cpu
[pip3] torchaudio==2.5.1+cu124
[pip3] torchmetrics==1.2.1
[pip3] torchtune==0.3.1
[pip3] torchvision==0.20.1+cu124
[pip3] triton==3.1.0
[conda] Could not collect

cc @msaroufim @chauhang @penguinwu

**Comment by zou3519** [comment]
Try:
```
bench()

a = datetime.now()
final_total = bench()
b = datetime.now()
```

Otherwise, you're benchmarking compilation time as well. After a warmup bench operations, torch.compile is much faster compared with no torch.compile
